### DISCUSSION TITLE
Trying to setup k3s in HA with external etcd.

### DISCUSSION THREAD

**Discussion by politick** [k3s-io/k3s#6912] [Q&A] [answered]
Hi there,
all my 5 k3s-server hang and never even listens on :6443

I have installed 5 etcd VMs running on 5 different nodes.
I have installed 5 k3s-server VMs on 5 different nodes. ( separate VMs from etcd )
I have verified that I can read any etcd from any of the k3s server with:
`etcdctl --cert-file /data/etcd/certs/etcd-client.politick.ca.pem  --key-file /data/etcd/certs/etcd-client.politick.ca-key.pem  --ca-file /data/etcd/certs/etcd-root-ca.pem  --endpoints https://172.16.9.35:2379  ls`

The k3s systemd install has created a file in `/etc/systemd/system.k3s.service.env` with this content:
```
K3S_DATASTORE_CAFILE='/data/etcd/certs/etcd-root-ca.pem'
K3S_DATASTORE_CERTFILE='/data/etcd/certs/etcd-client.politick.ca.pem'
K3S_DATASTORE_ENDPOINT='https://etcd-1.politick.ca:2379,https://etcd-2.politick.ca:2379,https://etcd-3.politick.ca:2379,https://etcd-4.politick.ca:2379,https://etcd-5.politick.ca:2379'
K3S_DATASTORE_KEYFILE='/data/etcd/certs/etcd-client.politick.ca-key.pem'
K3S_TOKEN='99ce9ce25d7215809127a4626be8f63e'
K3S_URL='https://k3s.politick.ca:6443'
```
and `/etc/systemd/system.k3s.service` with this content:
```[Unit]
Description=Lightweight Kubernetes
Wants=network-online.target
After=network-online.target
<SNIP> 
TimeoutStartSec=0
Restart=always
RestartSec=5s
ExecStartPre=/bin/sh -xc '! /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service'
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \
   server \
```

all the VMs have 2 NICs, on 2 separate networks (192.168.10.0/24  and 172.16.10.0/14)  I have the etcd configured to listen on the 172.16.10.0/14 address range, and the k3s on the 192.168.10.0/24 range because I started to run out of IP addresses on the /24)

When I issue a `systemctl start k3s`, the command NEVER RETURNS from systemctl and remains stuck for ever, I need to `<CTRL-C>` out of it, but k3s is still running after the ctrl-c.  There is close to nothing in the k3s logs:

```
root@k3ss-51:/var/log# journalctl -ru k3s
Feb 05 21:33:46 k3ss-51 k3s[9082]: time="2023-02-05T21:33:46Z" level=info msg="Starting k3s v1.25.6+k3s1 (9176e03c)"
Feb 05 21:33:45 k3ss-51 sh[9077]: Failed to get unit file state for nm-cloud-setup.service: No such file or directory
Feb 05 21:33:45 k3ss-51 sh[9076]: + /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
Feb 05 21:33:45 k3ss-51 systemd[1]: Starting Lightweight Kubernetes...
```

If I start it manually in a shell with debug, it will listen on :6443
k3s --debug has a lot of output but this might be of interest:
`time="2023-02-07T05:02:58Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:6443/v1-k3s/readyz: 500 Internal Server Error"`
but the problem when started manually is that the k3s servers don't create a cluster, they are all individuals.

I've not configured anything for the CNI, I assume that k3s takes care of that.  I've even tried to add `--flannel-external-ip` to the k3s command line, but it did not accept it.  I'm not even sure where to configure the CNI (if I need to ..)

but check config seem to return PASS
```
k3s check-config

Verifying binaries in /var/lib/rancher/k3s/data/630c40ff866a3db218a952ebd4fd2a5cfe1543a1a467e738cb46a2ad4012d6f1/bin:
- sha256sum: good
- links: good

System:
- /usr/sbin iptables v1.8.7 (nf_tables): ok
- swap: disabled
- routes: ok

Limits:
- /proc/sys/kernel/keys/root_maxkeys: 1000000

modprobe: FATAL: Module configs not found in directory /lib/modules/5.15.0-57-generic
info: reading kernel config from /boot/config-5.15.0-57-generic ...


info: reading kernel config from /boot/config-5.15.0-57-generic ...

Generally Necessary:
- cgroup hierarchy: cgroups V2 mounted, cpu|cpuset|memory controllers status: good
- /usr/sbin/apparmor_parser
apparmor: enabled and tools installed
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MEMCG: enabled
- CONFIG_KEYS: enabled
- CONFIG_VETH: enabled (as module)
- CONFIG_BRIDGE: enabled (as module)
- CONFIG_BRIDGE_NETFILTER: enabled (as module)
- CONFIG_IP_NF_FILTER: enabled (as module)
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_IPVS: enabled (as module)
- CONFIG_IP_NF_NAT: enabled (as module)
- CONFIG_NF_NAT: enabled (as module)
- CONFIG_POSIX_MQUEUE: enabled

Optional Features:
- CONFIG_USER_NS: enabled
- CONFIG_SECCOMP: enabled
- CONFIG_CGROUP_PIDS: enabled
- CONFIG_BLK_CGROUP: enabled
- CONFIG_BLK_DEV_THROTTLING: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CGROUP_HUGETLB: enabled
- CONFIG_NET_CLS_CGROUP: enabled (as module)
- CONFIG_CGROUP_NET_PRIO: enabled
- CONFIG_CFS_BANDWIDTH: enabled
- CONFIG_FAIR_GROUP_SCHED: enabled
- CONFIG_RT_GROUP_SCHED: missing
- CONFIG_IP_NF_TARGET_REDIRECT: enabled (as module)
- CONFIG_IP_SET: enabled (as module)
- CONFIG_IP_VS: enabled (as module)
- CONFIG_IP_VS_NFCT: enabled
- CONFIG_IP_VS_PROTO_TCP: enabled
- CONFIG_IP_VS_PROTO_UDP: enabled
- CONFIG_IP_VS_RR: enabled (as module)
- CONFIG_EXT4_FS: enabled
- CONFIG_EXT4_FS_POSIX_ACL: enabled
- CONFIG_EXT4_FS_SECURITY: enabled
- Network Drivers:
  - "overlay":
    - CONFIG_VXLAN: enabled (as module)
      Optional (for encrypted networks):
      - CONFIG_CRYPTO: enabled
      - CONFIG_CRYPTO_AEAD: enabled
      - CONFIG_CRYPTO_GCM: enabled
      - CONFIG_CRYPTO_SEQIV: enabled
      - CONFIG_CRYPTO_GHASH: enabled
      - CONFIG_XFRM: enabled
      - CONFIG_XFRM_USER: enabled (as module)
      - CONFIG_XFRM_ALGO: enabled (as module)
      - CONFIG_INET_ESP: enabled (as module)
      - CONFIG_INET_XFRM_MODE_TRANSPORT: missing
- Storage Drivers:
  - "overlay":
    - CONFIG_OVERLAY_FS: enabled (as module)

STATUS: pass
```

I don't know why the k3s server just hang.

Thank  for the help & Kind Regards
Martin Politick, Feb 2023.

**Comment by brandond** [comment, marked as answer]
You've specified a server address and token, which tells k3s that you want it to join an existing embedded etcd cluster. Don't specify K3S_URL on server nodes if you don't want to join an embedded etcd cluster.

**Reply by politick** [reply]
I'll try that.  
I thought the K3S_URL was : `Server to connect to, used to join a cluster`

**Reply by brandond** [reply]
Yes, for servers, that argument is the server to connect to when joining an embedded etcd cluster. You're not doing that, you're using an external datastore. K3s will retrieve all the information it needs using the datastore connection info and the token. You should be referencing these docs: https://docs.k3s.io/installation/ha

**Reply by politick** [reply]
I've just re-built all the VMs from scratch with Ansible (on Proxmox)
then installed k3s with Ansible without the K3S_URL.
but with the same effect.

New content of  `cat /etc/systemd/system/k3s.service.env`
```
K3S_DATASTORE_CAFILE='/data/etcd/certs/etcd-root-ca.pem'
K3S_DATASTORE_CERTFILE='/data/etcd/certs/etcd-client.politick.ca.pem'
K3S_DATASTORE_ENDPOINT='https://etcd-1.politick.ca:2379,https://etcd-2.politick.ca:2379,https://etcd-3.politick.ca:2379,https://etcd-4.politick.ca:2379,https://etcd-5.politick.ca:2379'
K3S_DATASTORE_KEYFILE='/data/etcd/certs/etcd-client.politick.ca-key.pem'
K3S_TOKEN='99ce9c09127a4626bee25d721588f63e'
```

I will try again to verify the etcd connection from all my new VMs, but I don't foresee that they would not just connect.

Kind Regards,
Martin Politick.

**Reply by politick** [reply]
On one of the new k3s-server VMs, I've verified that I can still reach the etcd with 
```
 . /etc/systemd/system/k3s.service.env
etcdctl --cert-file $K3S_DATASTORE_CERTFILE  --key-file $K3S_DATASTORE_KEYFILE  --ca-file $K3S_DATASTORE_CAFILE  --endpoints https://172.16.9.35:2379  ls
```

Oh, I think I just figured it out... I was using the IP address of etcd, not the name.
```
etcdctl --cert-file $K3S_DATASTORE_CERTFILE  --key-file $K3S_DATASTORE_KEYFILE  --ca-file $K3S_DATASTORE_CAFILE  --endpoints https://etcd-1.politick.ca:2379  ls
Error:  client: etcd cluster is unavailable or misconfigured; error #0: x509: certificate is valid for localhost, etcd-1, etcd, not etcd-1.politick.ca

error #0: x509: certificate is valid for localhost, etcd-1, etcd, not etcd-1.politick.ca

```

So I need to add the full name in the certificate maybe ?
I thought I did.  I'll investigate further and update this thread if that's it !!!

Thx !

Kind Regards,
Martin Politick.

**Reply by politick** [reply]
I was expecting that :
`"CN": "etcd-1.politick.ca",` was the canonical name for the machine for the cert.
I'll add it to the host[] list and try again.
```
{
  "key": {
    "algo": "rsa",
    "size": 2048
  },
  "names": [
    {
      "O": "Politick",
      "OU": "Politick K3s",
      "L": "Vancouver",
      "ST": "BC",
      "C": "CANADA"
    }
  ],
"CN": "etcd-1.politick.ca",
  "hosts": [
    "127.0.0.1",
    "localhost",
    "172.16.9.31",
    "etcd-1",
    "etcd"
  ]
}
``

**Reply by brandond** [reply]
The CN doesn't really matter any more for server certificates, you should make sure you have all the IP addresses and hostname variations in the host list that is used to populate the SANs on the cert.

**Reply by politick** [reply]
Yep !!!!

So 1 cfssl script modified,
git commit to my GitLab,
4 ansible playbooks and
25 new VMs ...
It's finally a go!

```
k3s kubectl get nodes
NAME      STATUS   ROLES                  AGE     VERSION
k3ss-11   Ready    control-plane,master   2m27s   v1.25.6+k3s1
k3ss-21   Ready    control-plane,master   119s    v1.25.6+k3s1
k3ss-31   Ready    control-plane,master   62s     v1.25.6+k3s1
k3ss-41   Ready    control-plane,master   22s     v1.25.6+k3s1
k3ss-51   Ready    control-plane,master   2m59s   v1.25.6+k3s1
```

But I must say that the way k3s fails on a bad cert for etcd is not what I'd expect...
I would have expected a bit more logs than just hang there ...

Thx  [Brad Davidson](https://github.com/brandond)

Kind Regards,
Martin Politick, February 2023
